pandas/core/indexing.py

```diff
@@ -2063,7 +2063,16 @@ def _convert_key(self, key, is_setter: bool = False):
 
         return key
 
+    @property
+    def _axes_are_unique(self) -> bool:
+        # Only relevant for self.ndim == 2
+        return self.obj.index.is_unique and self.obj.columns.is_unique
+
     def __getitem__(self, key):
+        if self.ndim == 2 and not self._axes_are_unique:
+            # GH#33041 fall back to .loc
+            return self.obj.loc[key]
```

Comment:
I know this is not the behaviour of `Series.at`, but the other option is also to raise an error if the requested key is duplicated. This way we can ensure it is always a scalar that is being returned.

Reply:
I was leaning that way until I saw that Series.at explicitly handles duplicate indexes

Reply:
I am personally in favor of raising, now we have the option to choose that.

Reply:
I see the appeal, but we shouldn't have the DataFrame behavior differ from the Series behavior.  So we would need to deprecate the Series behavior, right?

Also, this currently _doesnt_ fail if we have non-unique index/columns but are doing a lookup on a key that only occurs once, so raising for all non-unique cases would be a breaking change

Reply:
> Also, this currently doesnt fail if we have non-unique index/columns but are doing a lookup on a key that only occurs once

Can you give an example? I was trying it on the example from the test with an extra column, and that still fails:

```
arr = np.random.randn(9).reshape(3, 3) 
df = DataFrame(arr, columns=["A", "A", "B"])  
df.at[0, 'B']
```

Now, my proposal was actually also to only raise for a *duplicated key* (if that is possible to implement), so that would not be a breaking change then?

Reply:
> Now, my proposal was actually also to only raise for a duplicated key (if that is possible to implement), so that would not be a breaking change then?

Correct.  That would amount to a try/except in DataFrame._set_value and DataFrame._get_value with a better exception message re-raised on exception, pretty much at the exact places where #33052 adds assertions for self.columns.is_unique

Reply:
-1 on changing this, IOW I prefer to handle the non-unique as we do now (and do in .loc). If we change this we would need to deprecate all uses of this and would take quite a while. Making .at behave *the same* for now is prefereable.

Reply:
I would be fine with deprecating it for Series.at. 
(and I think it is fine to only consider `at` for now, as this is a more restricted indexer anyway compared to `.loc`, which needs to handle all kinds of different cases)

Reply:
I'm not wild about deprecating, would prefer to make that discussion separate.

Reply:
I opened https://github.com/pandas-dev/pandas/issues/33153 to have this discussion properly.